### PR TITLE
Implemented most methods on DescriptorPool

### PR DESCRIPTION
**Pull request by haberman**
This fixes most of the errors we were seeing in `setUp()` functions on test classes, which allows us to remove most of the test skips we were doing before.

### CODE REVIEW

File: python/descriptor.c [hunk 1]
```diff
@@ -144,34 +145,33 @@ static PyObject* PyUpb_DescriptorBase_GetOptions(PyUpb_DescriptorBase* self,
...
-  Py_XDECREF(str);
+  PyObject* str = PyBytes_FromStringAndSize(pb, size);
   upb_arena_free(arena);
-  return ret;
+  return str;
 
 oom:
   PyErr_SetNone(PyExc_MemoryError);
-  goto done;
+  return NULL;
```
Comment (fowles) [inline python/descriptor.c:164]: pretty sure you are missing an arena free call on the oom path

File: python/descriptor.c [hunk 2]
```diff
@@ -1065,6 +1065,13 @@ static PyObject* PyUpb_FileDescriptor_GetPackage(PyObject* _self,
   return PyUnicode_FromString(upb_filedef_package(self->def));
 }
 
+static PyObject* PyUpb_FileDescriptor_GetSerializedPb(PyObject* self,
+                                                      void* closure) {
```
Comment (fowles) [inline python/descriptor.c:1069]: do we not use the closure object?

File: python/descriptor_pool.c [hunk 3]
```diff
@@ -184,40 +219,195 @@ static PyObject* PyUpb_DescriptorPool_FindExtensionByName(PyObject* _self,
...
+
+  const char* last_dot = strrchr(name, '.');
+  if (last_dot) {
+    const upb_msgdef* parent =
+        upb_symtab_lookupmsg2(self->symtab, name, last_dot - name);
+    if (parent) {
+      const upb_oneofdef* o = upb_msgdef_ntooz(parent, last_dot + 1);
+      return PyUpb_OneofDescriptor_Get(o);
+    }
+  }
```
Comment (fowles) [inline python/descriptor_pool.c:317]: looks like the second time you have this strchr logic.  Can you share code?

File: python/descriptor_pool.c [hunk 4]
```diff
@@ -184,40 +219,195 @@ static PyObject* PyUpb_DescriptorPool_FindExtensionByName(PyObject* _self,
...
+
+static PyObject* PyUpb_DescriptorPool_FindAllExtensions(PyObject* _self,
+                                                        PyObject* msg_desc) {
+  PyUpb_DescriptorPool* self = (PyUpb_DescriptorPool*)_self;
+  const upb_msgdef* m = PyUpb_Descriptor_GetDef(msg_desc);
+  size_t n;
+  const upb_fielddef** ext = upb_symtab_getallexts(self->symtab, m, &n);
+  PyObject* ret = PyList_New(n);
+  for (size_t i = 0; i < n; i++) {
+    PyObject* field = PyUpb_FieldDescriptor_Get(ext[i]);
```
Comment (fowles) [inline python/descriptor_pool.c:378]: can this fail?

File: upb/util/def_to_proto.c [hunk 5]
```diff
@@ -392,10 +392,14 @@ static google_protobuf_FileDescriptorProto *filedef_toproto(
   google_protobuf_FileDescriptorProto_set_name(
       proto, strviewdup(ctx, upb_filedef_name(f)));
 
-  size_t n = strlen(upb_filedef_package(f));
-  if (n) {
-    google_protobuf_FileDescriptorProto_set_package(
-        proto, strviewdup(ctx, upb_filedef_package(f)));
+  const char* package = upb_filedef_package(f);
+  size_t n;
```
Comment (fowles) [inline upb/util/def_to_proto.c:396]: the value of n from this block is not used, so I think you should embed the local into the block to keep its scope tighter

**Review by fowles — APPROVED** [review]

File: python/descriptor.c (hunk 1 above)
Comment (haberman) [reply python/descriptor.c:164]: Done.

File: python/descriptor.c (hunk 2 above)
Comment (haberman) [reply python/descriptor.c:1069]: No we don't generally use it ever.

File: python/descriptor_pool.c (hunk 3 above)
Comment (haberman) [reply python/descriptor_pool.c:317]: Done.

File: python/descriptor_pool.c (hunk 4 above)
Comment (haberman) [reply python/descriptor_pool.c:378]: For OOM probably. Added a check.

File: upb/util/def_to_proto.c (hunk 5 above)
Comment (haberman) [reply upb/util/def_to_proto.c:396]: Done. `n` is also used later in the function, but I moved the declaration down to the next usage.

**Pull request merged**